replacing node module with externals fails (Module not found: TypeError: Cannot read property 'test' of undefined)

# Bug report

I get the following error when running the newest version of webpack `v4.28.0`.

```
Module not found: TypeError: Cannot read property 'test' of undefined
```

It happens whenever I try and import a dependency that only works on Node.JS that I've set to an empty object (for example `fs`).

Here is my webpack externals config.

```js
webpackConfig.externals = {
  child_process: {},
  deasync: {},
  fs: {},
  winston: {}
}
```

Please note that this works fine on webpack `v4.27.1`.

Here is the full stack trace.

```
ERROR in ./node_modules/@reactant/core/lib/log/winston.js
Module not found: TypeError: Cannot read property 'test' of undefined
 @ ./node_modules/@reactant/core/lib/log/winston.js 17:15-33
 @ ./node_modules/@reactant/core/lib/log/index.js
 @ ./node_modules/@reactant/core/lib/index.js
 @ /Users/codejamninja/Projects/reactant/packages/web/lib/hotDevClient.js
 @ multi (webpack)-dev-server/client?http://0.0.0.0:6002 (webpack)/hot/dev-server.js ./web/client.js /Users/codejamninja/Projects/reactant/packages/web/lib/hotDevClient.js
```

Instead of throwing an error, it should simply require the module as an empty object.
Because the error is happening inside of webpack, the stack trace does not tell me what is going wrong. I've searched all over for this supposed 'test' property. I am not using a property called test anywhere in my code.

This also fails with other node-specific npm modules, such as `winston` and `deasync`.

**Other relevant information:**
webpack version:  
Node.js version: v8.14.0
Also tested on Node.js version: v10.14.
Operating System: Ubuntu 18.10

I have the same problem since yesterday. The error seems to be in "require.resolve":

```
ERROR in ../node_modules/summernote/dist/summernote.js
Module not found: TypeError: Cannot read property 'test' of undefined
 @ ../node_modules/summernote/dist/summernote.js 664:14-43 4749:10-4751:12
 @ ./src/libs.ts
 @ ./index.ts
```

The line 664 (marked with <---) in summernote:
```
      if (typeof __webpack_require__ === 'function') { // eslint-disable-line
          try {
              // If CodeMirror can't be resolved, `require.resolve` will throw an
              // exception and `hasCodeMirror` won't be set to `true`.
              require.resolve('codemirror');                                    // <--- Line 664
              hasCodeMirror = true;
          }
          catch (e) {
              // do nothing
          }
      }
```

In my `webpack.config.js`, I have set:
```
externals: {
    codemirror: {}
},
```
and
```
new webpack.IgnorePlugin(/^codemirror$/),
```

I'm having probably related issue after upgrading to `4.28.0`, build output gives me:

```
WARNING in ./node_modules/moment/moment.js
Module not found: TypeError: Cannot read property 'request' of null
```

i'm using `webpack.IgnorePlugin` as documented here:
https://webpack.js.org/plugins/ignore-plugin/#example-of-ignoring-moment-locales

Either
```js
    new webpack.IgnorePlugin({
      resourceRegExp: /^\.\/locale$/,
      contextRegExp: /moment$/
    }),
```
or 
```
new webpack.IgnorePlugin({/^\.\/locale$/, /moment$/)
```

breaks with the latest release.

--- a/lib/IgnorePlugin.js
+++ b/lib/IgnorePlugin.js
@@ -40,10 +40,11 @@ class IgnorePlugin {
 	checkIgnore(result) {
 		if (
 			"checkResource" in this.options &&
+			this.options.checkResource &&
 			this.options.checkResource(result.request, result.context)
 		) {
 			// TODO webpack 5 remove checkContext, as checkResource already gets context
-			if ("checkContext" in this.options) {
+			if ("checkContext" in this.options && this.options.checkContext) {
 				if (this.options.checkContext(result.context)) {
 					return null;
 				}
@@ -54,9 +55,10 @@ class IgnorePlugin {
 
 		if (
 			"resourceRegExp" in this.options &&
+			this.options.resourceRegExp &&
 			this.options.resourceRegExp.test(result.request)
 		) {
-			if ("contextRegExp" in this.options) {
+			if ("contextRegExp" in this.options && this.options.contextRegExp) {
 				// if "contextRegExp" is given,
 				// both the "resourceRegExp" and "contextRegExp" have to match.
 				if (this.options.contextRegExp.test(result.context)) {

--- a/test/configCases/errors/multi-entry-missing-module/webpack.config.js
+++ b/test/configCases/errors/multi-entry-missing-module/webpack.config.js
@@ -8,9 +8,7 @@ module.exports = {
 		filename: "[name].js"
 	},
 	plugins: [
-		new IgnorePlugin({
-			resourceRegExp: new RegExp(/intentionally-missing-module/)
-		})
+		new IgnorePlugin(/intentionally-missing-module/)
 	],
 	node: {
 		__dirname: false

